gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/service/granter/extensiongrant/ExtensionGrantGranter.java

```diff
@@ -124,6 +128,7 @@ public boolean handle(String grantType, Client client) {
                                         user.setRoles(idpUser.getRoles());
                                         return user;
                                     })
+                                    .flatMap(user -> userService.enhance(user).toMaybe())
```

Comment:
Hey @micivray, you cannot use directly the `enhance` method because the current user hold the ID of the IDP user (external_id). You need first to retrieve the internal user from Gravitee, here is a snippet

```
.flatMap(user -> userService.findByDomainAndExternalIdAndSource(domain, user.getId(), idp)
                                 .flatMap(userService.enhance(user).toMaybe())
                                 .defaultEmpty(user))

Reply:
Ok thank you @tcompiegne. I didn't think that I needed to load any other information as it was working as is (local test with a LDAP IDP) just setting: user id with idpUser id.

In this case do you advice to call userService.findByDomainAndUsernameAndSource if userService.findByDomainAndExternalIdAndSource is not found (empty) ?

Reply:
Hi TItouan!
Here is a little update if this can help understand why @micivray is struggling with the suggestion you made.
Our flow is: 
(end user / RO) --> Client App --> Corporate OIDC server --> get an ID_TOKEN, then used a JWT bearer --> (Gravitee APIM) --> custom API inserting the user identity into an OpenLDAP, within a group carried by one of the JWT Bearer claim --> Gravitee AM --> OpenLDAP (as IdP) to fetch the user's group --> Gravitee AM to map user's group to a role and to dynamically add scopes

We have a more detailed procedure (24 pages) on how-to reproduce the bug detailed in issue https://github.com/gravitee-io/issues/issues/3534, that we can share privately.

So here, as @micivray said, the "user is not saved in our case", as you expect in your suggestion.

What do you think?

Reply:
You can see my proposal with the fix below to enhance user roles and permissions without calling the enhance method with the wrong ID.